### DISCUSSION TITLE
typealias でプラットフォーム間の名前の差異を吸収する。

### DISCUSSION THREAD

**Discussion by YusukeHosonuma** [YusukeHosonuma/Effective-SwiftUI#17] [Ideas]
## Overview

UIKit と AppKit は API 設計の思想が似ているため、似たような名前で対称となるクラスがいくつか存在する。例えば、`UIView` / `NSView`、`UIColor` / `NSColor`、`UIImage` / `NSImage`、など。

それらのメソッドやプロパティといった内部的なデータにアクセスする必要がなく、コンパイル時に型名だけが必要となる場合は `typealias` で共通の名前をつけることによって、コードの記述量を減らすことができるケースがある。

```swift
#if canImport(AppKit)
import AppKit
typealias XView = NSView // 💡 `XView` is a `NSView` on macOS
#endif

#if canImport(UIKit)
import UIKit
typealias XView = UIView // 💡 `XView` is a `UIView` on iOS
#endif
```

## Discussion

より身近な例として [UIViewRepresentable](https://developer.apple.com/documentation/swiftui/uiviewrepresentable) / [NSViewRepresentable](https://developer.apple.com/documentation/swiftui/nsviewrepresentable) の両方を実装する必要があるときにも利用できる。

```swift
#if os(macOS)
typealias Representable = NSViewRepresentable
#else
typealias Representable = UIViewRepresentable
#endif

public struct XxxView: Representable { ... }
```

ただし、プロトコルのメソッド名についても `makeUIView` / `makeNSView `、`updateUIView` / `updateNSView` といった違いがあるので、内部的に private メソッドなどを用意し、それを共通的に呼ぶようにするとよい。

```swift
public struct XxxView: Representable {

    // 💡 Adopt to UIViewRepresentable.
    public func makeUIView(context: Context) -> WKWebView {
        makeXxxView(context: context)
    }

    // 💡 Adopt to NSViewRepresentable.
    public func makeNSView(context: Context) -> WKWebView {
        makeXxxView(context: context)
    }

    // ✅ Common method.
    private func makeWKWebView(context: Context) -> XxxView { ... }

    ...
}
```

## Why not?

`UIxxx` と `NSxxx` は名前こそ似ているが、それぞれのメソッドやプロパティなどに差異は存在する。そのため、内部データにアクセスする必要がある場合は、こうした共通化の仕組みは却って足かせになるケースも考えられる。利用する場合はユースケースとして適切であるか検討すること。

## Example

[WKWebKit を UIViewRepresentable /  NSViewRepresentable でラップした実装例](https://github.com/YusukeHosonuma/SwiftUICommon/blob/main/Sources/SwiftUICommon/View/WebView.swift)

## Reference

- [UIViewRepresentable](https://developer.apple.com/documentation/swiftui/uiviewrepresentable)
- [NSViewRepresentable](https://developer.apple.com/documentation/swiftui/nsviewrepresentable)

**Comment by zunda-pixel** [comment]
Apple公式サンプルでも同様の手段を取っていたので、載せておきます。

https://github.com/apple/sample-food-truck/blob/main/App/City/DetailedMapView.swift

**Reply by uhooi** [reply]
おおおホントだ…！
トップレベルに定義するときは `private` を付けているのもいいですね…！
https://github.com/apple/sample-food-truck/blob/f65b77090ee990634a5140f97f2adae95945b041/App/City/DetailedMapView.swift#L11-L15

**Reply by uhooi** [reply]
> ```swift
> #if os(iOS)
> private typealias ViewControllerRepresentable = UIViewControllerRepresentable
> #elseif os(macOS)
> private typealias ViewControllerRepresentable = NSViewControllerRepresentable
> #endif
> ```
